Reveal Pre-images beyond the end of the enrollment cycle

hewison-chris opened the pull request:
Reveal pre-images beyond the current cycle. This will help reduce the risk of being slashed when the new cycle starts.

hewison-chris committed:
Trivial removal of unused function

hewison-chris committed:
ValidatorRecurringEnrollment test with 3 enrolled

This test truncates the 6 enrollments in block 20, so that only 3 nodes
are enrolled in block 20, should only expect the first three nodes to
send preimages for block 21.

linked0 commented:
Good try! But I don't think this should solve the root cause because what you're intending to do in the PR is already done as follows in Base.d
```
        // Check block is at target height for the participating clients
        const enroll_block = first_client.getBlock(enrolled_height);
        this.expectHeightAndPreImg(client_idxs, target_height,
            enroll_block.header, 10.seconds, file, line);
```
It happens fifty-fifty in local machines, which means to be kind of tricky. We could raise an issue and handle it.

hewison-chris committed:
Test.base.d correction to waitForPreimages logging

hewison-chris committed:
Some trivial debug logging

hewison-chris committed:
Allow preimage reveal beyond enrollemnt

If the pre-images are only revealed up to the end of the enrollment
cycle then there is more risk for the validator to get slashed in the
first block of the new cycle.

hewison-chris commented:
> Good try! But I don't think this should solve the root cause because what you're intending to do in the PR is already done as follows in Base.d
> 
> ```
>         // Check block is at target height for the participating clients
>         const enroll_block = first_client.getBlock(enrolled_height);
>         this.expectHeightAndPreImg(client_idxs, target_height,
>             enroll_block.header, 10.seconds, file, line);
> ```

But the `enrolled_height` here is assumed to be at the end of the cycle which is not the case for the test that limits the `enrollments` to 3 in block 20 and has then has 3 in block 21. That is why we should not rely on this function for the test which does not follow the expected pattern of enrollment.

linked0 commented:
My comment was for the previous code that was, if I'm correct, passing in the enrollments for this function. But the code has been changed. Anyway, I'm thinking that the `expectHeightAndPreImg` has to have all the current enrollments for `target_height` as a parameter but it doesn't now, which I think you might have meant. I will fix the code in this sprint.

hewison-chris committed:
ValidatorRecurringEnrollment test: add assertion for enrolls

hewison-chris committed:
Trivial log correction for rejected preimage

hewison-chris committed:
Do not update stored preimage if we have higher

hewison-chris committed:
Do not reset validator enroll_key

hewison-chris committed:
Trivial removal of unused function

hewison-chris commented:
Closing as we have https://github.com/bosagora/agora/pull/3153
